crypto/x509/x_name.c

```diff
@@ -310,16 +309,18 @@ static int x509_name_canon(X509_NAME *a)
 
     OPENSSL_free(a->canon_enc);
     a->canon_enc = NULL;
-    /* Special case: empty X509_NAME => null encoding */
+
+    /* Special case: empty X509_NAME (NULL-DN) */
     if (sk_X509_NAME_ENTRY_num(a->entries) == 0) {
         a->canon_enclen = 0;
+        if ((a->canon_enc = (unsigned char *)OPENSSL_strdup("")) == NULL)
+            goto oom;
```

Comment:
I do not think that this is the right approach - we should not be allocating a pointer to a zero length item as a pattern in the code. This should be fixed via another approach.

The common oom error handling making sense however - but that is unrelated to this bug fix..

Reply:
My original plan was to use a static empty string, but this does not fix here because in all other (not NULL-DNI) cases `a->canon_enc` is dynamically allocated.

Why do you think we should not allocate a pointer in this case? To be slightly more efficient?
The downside of error-prone handling of the special NULL case would weigh more here,
as can be seen from the issue being fixed.